error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/Slf4jLogDeclaration.java

```diff
@@ -0,0 +1,129 @@
+import com.sun.source.tree.VariableTree;
+import javax.inject.Inject;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.Name;
+import javax.tools.JavaFileObject;
+import tech.picnic.errorprone.bugpatterns.util.SourceCode;
+
+/**
+ * A {@link BugChecker} that warns when SLF4J declarations are not canonicalized across the project.
+ *
+ * @apiNote The default canonicalized logger name can be overriden through {@link ErrorProneFlags
+ *     flag arguments}.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Make sure SLF4J log declarations follow the SLF4J conventions inside classes.",
+    link = BUG_PATTERNS_BASE_URL + "Slf4jLogDeclaration",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = LIKELY_ERROR)
+public final class Slf4jLogDeclaration extends BugChecker
+    implements ClassTreeMatcher, MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Matcher<ClassTree> TEST_CLASS_WITH_LOGGER =
+      allOf(hasIdentifier(isSubtypeOf("org.slf4j.Logger")));
+  private static final Matcher<Tree> LOGGER = isSubtypeOf("org.slf4j.Logger");
+  private static final String CANONICALIZED_LOGGER_NAME_FLAG =
+      "Slf4jLogDeclaration:CanonicalizedLoggerName";
+  private static final String DEFAULT_CANONICALIZED_LOGGER_NAME = "LOG";
+  private static final Matcher<ExpressionTree> GET_LOGGER_METHOD =
+      staticMethod().onDescendantOf("org.slf4j.LoggerFactory").named("getLogger");
+
+  private final String canonicalizedLoggerName;
+
+  /** Instantiates a default {@link Slf4jLogDeclaration} instance. */
+  public Slf4jLogDeclaration() {
+    this(ErrorProneFlags.empty());
+  }
+
+  /**
+   * Instantiates a customized {@link Slf4jLogDeclaration}.
+   *
+   * @param flags Any provided command line flags.
+   */
+  @Inject
+  Slf4jLogDeclaration(ErrorProneFlags flags) {
+    canonicalizedLoggerName = getCanonicalizedLoggerName(flags);
+  }
+
+  @Override
+  public Description matchClass(ClassTree tree, VisitorState state) {
+    if (tree.getKind() == Kind.INTERFACE || !TEST_CLASS_WITH_LOGGER.matches(tree, state)) {
+      return Description.NO_MATCH;
+    }
+
+    SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+    for (Tree member : tree.getMembers()) {
+      if (LOGGER.matches(member, state)) {
+        VariableTree variable = (VariableTree) member;
+        SuggestedFixes.addModifiers(
+                member, state, Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
+            .ifPresent(fixBuilder::merge);
+        if (!variable.getName().toString().startsWith(canonicalizedLoggerName)) {
+          fixBuilder.merge(SuggestedFixes.renameVariable(variable, canonicalizedLoggerName, state));
+        }
+      }
+    }
+
+    return describeMatch(tree, fixBuilder.build());
+  }
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    if (!GET_LOGGER_METHOD.matches(tree, state)) {
```

Comment:
We're now creating separate fixes for the field declaration (modifiers + name) and for the referenced class. I wonder whether this means that we should have two _separate_ checks. I can imagine that some users will appreciate the second, while finding the first too conservative.

Reply:
Pushed a change using TreeScanner (Conceptually still two checks using`visitClass` & `visitMethodInvocationTree`).

What do you think about making the separation into an EP flag ?